src/evo/cbtx.cpp

```diff
@@ -314,8 +322,139 @@ bool CalcCbTxMerkleRootQuorums(const CBlock& block, const CBlockIndex* pindexPre
     return true;
 }
 
+bool CheckCbTxBestChainlock(const CBlock& block, const CBlockIndex* pindexPrev, const llmq::CChainLocksHandler& chainlock_handler, BlockValidationState& state)
+{
+    if (block.vtx[0]->nType != TRANSACTION_COINBASE) {
+        return true;
+    }
+
+    CCbTx cbTx;
+    if (!GetTxPayload(*block.vtx[0], cbTx)) {
+        return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-cbtx-payload");
+    }
+
+    if (cbTx.nVersion >= CCbTx::CB_V20_VERSION) {
+        auto prevBlockCoinbaseChainlock = GetNonNullCoinbaseChainlock(pindexPrev);
+        // If std::optional prevBlockCoinbaseChainlock is empty, then up to the previous block, coinbase Chainlock is null.
+        if (prevBlockCoinbaseChainlock.has_value()) {
+            // Previous block Coinbase has a non-null Chainlock: current block's Chainlock must be non-null and at least as new as the previous one
+            if (!cbTx.bestCLSignature.IsValid()) {
+                // IsNull() doesn't exist for CBLSSignature: we assume that a non valid BLS sig is null
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-cbtx-clsig");
+            }
+            int prevBlockCoinbaseCLHeight = pindexPrev->nHeight - static_cast<int>(prevBlockCoinbaseChainlock.value().second) - 1;
+            int curBlockCoinbaseCLHeight = pindexPrev->nHeight - static_cast<int>(cbTx.bestCLHeightDiff) - 1;
+            if (curBlockCoinbaseCLHeight < prevBlockCoinbaseCLHeight) {
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-cbtx-clsig");
+            }
+        }
+
+        // IsNull() doesn't exist for CBLSSignature: we assume that a valid BLS sig is non-null
+        if (cbTx.bestCLSignature.IsValid()) {
+            int curBlockCoinbaseCLHeight = pindexPrev->nHeight - static_cast<int>(cbTx.bestCLHeightDiff) - 1;
+            uint256 curBlockCoinbaseCLBlockHash = ::ChainActive()[curBlockCoinbaseCLHeight]->GetBlockHash();
```

Comment:
ChainActive locks cs_main; we should probably avoid that and do something like
```
            uint256 curBlockCoinbaseCLBlockHash = [&](){
                auto index = pindexPrev;
                for (auto i : irange::range(cbTx.bestCLHeightDiff))
                    index = index->pprev;
                return index->GetBlockHash();
            }();
```

although an even simpler solution looks like
```
            uint256 curBlockCoinbaseCLBlockHash = pindexPrev->GetAncestor(curBlockCoinbaseCLHeight)->GetBlockHash();
```